### PR TITLE
Add a custom mosaic + Update segmentation loss function

### PR DESCRIPTION
**Pull request by avv-va**
This PR creates two augment classes:
- RasterizeBoxes: creates a bounding box image label from bounding box coordinates such that the bounding box image highlights all the area that bounding box coordinates covers
- CustomMosaic: performs random slicing and shuffling per image

The loss function for PoseSeg task is updated such that:
- The segmentation loss is performed using the bounding box image instead of coordinates
- When a data is augmented through the custom mosaic, only segmentation loss is called

### CODE REVIEW

File: ultralytics/data/augment.py [hunk 1]
```diff
@@ -128,6 +130,52 @@ def get_indexes(self):
...
+class RasterizeBoxes:
+    def __call__(self, labels):
+        gt_labels = labels['instances']
+        img = labels['img']
+        h, w, _ = img.shape
+        gt_labels.convert_bbox(format='xyxy')
+        gt_labels.denormalize(w, h)
+        gt_bboxes = gt_labels.bboxes
+        gt_cls = labels['cls']
+        num_classes = int(gt_cls.max()) + 1 if len(gt_cls) != 0 else 2
```
[the remark this answers is not in the document]
Comment (avv-va) [reply ultralytics/data/augment.py:142]: When it was set to 0, I sometimes got this error during training: `/opt/conda/conda-bld/pytorch_1682343967769/work/aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [95,0,0], thread: [5,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.`  I couldn't find the exact cause but not setting it to 0 solved the issue. Should I find a way to set this to zero instead of the hack I am using?

File: ultralytics/utils/loss.py [hunk 2]
```diff
@@ -599,64 +561,88 @@ def __call__(self, preds, batch):
...
-            mask_gt (B, T, 1): mask_gt[b_idx][gt_idx] = 1.0 if gt exists else 0.0
-        Returns:
-            loss (torch.Tensor): calculated loss
+            pred_seg (B, A, seg_ch_num)
+            bboxes_img (B, C, A)
         """
+        # TODO: Should collapse bboxes_img if needed depending on C vs seg_ch_num
+        target_seg = gt_bboxes_img.permute(0, 2, 1) # B, A, C
+        loss_per_anchor = self.bce_inside(pred_seg, target_seg)  # (B, A, C)
+        anchor_weights = torch.ones_like(loss_per_anchor)  # (B, A, C)
```
[the remark this answers is not in the document]
Comment (romainthibaux) [reply ultralytics/utils/loss.py:593]: I would remove it. In general, it's almost always a bad idea to add code that is not useful now just because it might be useful later.
